lib/subscription-connection.js

```diff
@@ -52,7 +55,7 @@ module.exports = class SubscriptionConnection {
 
     switch (type) {
       case GQL_CONNECTION_INIT:
-        this.sendMessage(GQL_CONNECTION_ACK)
+        this.handleConnectionInit(data)
```

Comment:
```suggestion
        await this.handleConnectionInit(data)
```

Reply:
This might be harder than this to fix. Essentially it's possible to bypass the connectionInit hook by sending two messages without waiting for an ack. While we are doing connectionInit, we cannot process any more messages.

This feature is often called pipelining, and I'm not sure GraphQL subscriptions even supports that. A better fix would be to let the logic process only once message at a time.

Reply:
I tried to fix this and add some kind of pipelining inspired from on [Apollo implementation](https://github.com/apollographql/subscriptions-transport-ws/blob/52758bfba6190169a28078ecbafd2e457a2ff7a8/src/server.ts). 
See: https://github.com/PacoDu/fastify-gql/commit/df59562a02b7d8033b267a9d8dc31e07ef8b59e0 

But tests are inconsistent (sometime pass, sometime fail), I guess this is related to promises that are not properly handled, it would be nice to have your inputs on this one ! Also I think that the current version will allow `start` payload even if `connection_init` hasn't been sent.

Reply:
Yes, exactly. We need to stregthen up a bit the current implementation otherwise it would be fairly easy to overcome any auth check in `onConnect`.

I would recommend to turn the `socket` into a stream, and the use async iteration to process all messages one by one.

Reply:
Hi, sorry I didn't had much time to work on this last week. I'm not sure how to address this, could you please give me some details/leads on the changes it will imply for the message handler to turn the socket into a stream ? 

I guess I've to use this https://www.npmjs.com/package/ws#use-the-nodejs-streams-api but what about the message handler ? (I'm not confortable with stream yet - but I'm ready to try and learn)

Reply:
You can do something like:

```js
const stream = WebSocket.createWebSocketStream(ws, { encoding: 'utf8' });
for await (let msg of stream) {
  await handleMesasge(mg)
}
```

Reply:
😅ok pretty simple, thanks !

Reply:
I'm unable to catch any message from this loop. I'm passing the connection from the wsHandler to the SubscriptionConnection class and then create the stream and the async iteration loop in a `consume` method called from the wsHandler after the `new SubscriptionConnection`. I removed the `on('data')` handlers but the loop is never triggered. I'll try to find some times to investigate further next weekend but I think I'm missing something here, and I can't find what it is for now ...

Reply:
Can you push the broken code somewhere?

Reply:
https://github.com/PacoDu/fastify-gql/commit/fc8b823eb2b93dffb4d7a182685e99e6060c959d

Reply:
Do you have any idea of what I'm doing wrong here ? PacoDu@fc8b823

The issue here is that the `consume` method (https://github.com/PacoDu/fastify-gql/commit/fc8b823eb2b93dffb4d7a182685e99e6060c959d#diff-d0a65557fba21a77676fdc6c8ab721c7R47) is not receiving any message, it's called from the wsHandler here: https://github.com/PacoDu/fastify-gql/commit/fc8b823eb2b93dffb4d7a182685e99e6060c959d#diff-49234199fc0c7c90d47d4a9524c21f0bR41

Reply:
Resolved by #228